README
The Open Dynamics Engine (ODE), Copyright (C) 2001-2003 Russell L. Smith.
-------------------------------------------------------------------------

ODE is a free, industrial quality library for simulating articulated
rigid body dynamics - for example ground vehicles, legged creatures,
and moving objects in VR environments. It is fast, flexible, robust
and platform independent, with advanced joints, contact with friction,
and built-in collision detection.

This library is free software; you can redistribute it and/or
modify it under the terms of EITHER:
  (1) The GNU Lesser General Public License as published by the Free
      Software Foundation; either version 2.1 of the License, or (at
      your option) any later version. The text of the GNU Lesser
      General Public License is included with this library in the
      file LICENSE.TXT.
  (2) The BSD-style license that is included with this library in
      the file LICENSE-BSD.TXT.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the files
LICENSE.TXT and LICENSE-BSD.TXT for more details.

  * Installation instructions are in the INSTALL file

  * The ODE web pages are at http://q12.org/ode/

  * The ODE documentation is in the file ode/doc/ode.html, or you
    can view it on the web at http://q12.org/ode/


CHANGELOG
ODE CHANGELOG
-------------

the rules for this file:
  * entries are sorted newest-first.
  * summarize sets of changes - dont reproduce every CVS log comment here.
  * don't ever delete anything.
  * keep the format consistent (79 char width, M/D/Y date format).

------------------------------------------------------------------------------
02/04/06 gcarlton

	* Added support for geom offsets.

10/26/05 rodrigo

	* Removed LIBTOOL from autotools since it was not really required.
	* Added a target to build ODE as a shared library, this shared
	  library gets build alongside the static one, no flags required.

10/24/05 tfautre

	(Backported patches from STABLE branch, applied by Adam)
	
	* dRandInt changed for a non-double all-int version.
	* mics minor fixes and improvements.

04/05/05 tfautre

	* Fixed segmentation fault with OPCODE on 64 bits systems.

03/31/05 tfautre

	* Fixed timer.cpp compiler error on x86-64 using GCC.

03/29/05 colin

	* Added trimesh preprocessing to mark unneeded edges and verts.  Also
	  added support for preprocessed info to the ccylinder-trimesh
	  collider.

12/07/04 adam

        * Important AMotors bugfix

09/22/04 jeff

	* Assorted small bugfixes and tweaks for
          trimesh_{box,ccylinder,trimesh} collisions

09/21/04 jeff

	* added functions to joint.cpp to allow joint attachment to moving
	  geoms.

	* added malloc-based memory allocation in step.cpp & lcp.cpp (turned
	  on with a #define switch in common.h)
	
05/29/04 russ

	* added joint feedback to the QuickStep solver

05/18/04 russ

	* added warm starting to the QuickStep solver

05/18/04 russ

	* added the QuickStep solver
	
	* added contact parameter functions.

05/05/04 adam

        * use dRandInt instead of rand() in stepfast.

04/21/04 russ

	* added auto-disable support from Aras Pranckevicius (with
	  modifications by russ). this useful feature can speed up
	  simulation significantly in some cases.
	
	* various internal tidyups.

04/20/04 russ

	* changed the meaning of the 'index' argument to dJointGetBody():
	  it was the only remaining API function that does not respect
	  dJOINT_REVERSE (spotted by Matthew D. Hancher).
	
	* updated the C++ headers: fixed two minor bugs and added
	  support for dQuadTreeSpace, dRay, and the dGeom::getSpace() method
	  (from Matthew D. Hancher).

04/18/04 russ

	* changed the way that the dInfinity constant is implemented: now it
	  is #defined to be one of: FLT_MAX, DBL_MAX, HUGE_VAL, HUGE_VALF, or
	  a large numeric constant. previously it was a variable that was
	  exported from the library. this simplifies the configuration and
	  build process quite a bit, especially in the case of DLLs.
	
	* removed the old, deprecated collision system (geom.cpp,space.cpp,
	  geom.h,space.h,odecpp_old_collision.h). the ODE_OLD_COLLISION
	  configuration setting no longer has any meaning.
	
	* removed support for dGeomGroups, which have been deprecated for
	  a while and are equivalent to 'spaces' anyway.

04/13/04 russ

	* bug fix in dMassSetCappedCylinder(), from Matthew D. Hancher.

04/08/04 russ

	* added trimesh-CCylinder capability, from Vadim Macagon
	  <vadim_mcagon@hotmail.com>.

04/04/04 adam

        * yet another rewrite of triangle-box collision code, this
        time based on code donated by Croteam, ported by asko@jetti.org
        and tweaked by Erwin.

04/04/04 adam

        * merged trimesh-trimesh collision code by
        Jeffrey Smith <jeffreys@Softimage.com>.

        * changed it to not break the trimesh interface, fix
        some GCC compilation problems, bring it up to date with
        ODE changes from 2003-11-15 -> 2004-04-04.

        * add ability to drop meshes on meshes in test_moving_trimesh,
        not as good as it could be but it's illustrative.

01/16/04 adam

	* implement a bunch of ultra-simple TriMesh functions that were
	in the headers but not in the code -- patch by
	Vadim Macagon <vadim_mcagon@hotmail.com>

	* disable temporal coherence on trimeshes by default, since
	it has scaleability issues that don't make it a general clear win.

12/01/03 adam

	* implement dxHashSpace::collide2(), not particularly efficiently.

11/14/03 adam

	* applied several Trimesh fixes and improvements from
	Aras Pranckevicius <nearaz@interamotion.com>

10/22/03 adam

	* apply Nguyen Binh's work for removing many dSetZero() calls
	and some other extraneous initializations.

07/29/03 martin

	* added dJointAdd*Torque/Force().

07/10/03 russ

	* added the StepFast code, by David Whittaker.

07/02/03 martin

	* added dMassSet*Total().

07/01/03 martin

	* added joint limits and motors to universal joints.

	* reversed the polarity of the dJOINT_REVERSE flag.

06/30/03 russ

	* added the TriMesh geom class and the quad tree space to the ODE
	  core. both of these were developed by Erwin de Vries. added OPCODE
	  to the ODE distribution, this is required by TriMesh.

06/23/03 martin

	* added dGeomSetQuaternion() and dGeomGetQuaternion()
	
	* added dJointGet*Anchor2()

05/07/03 russ

	* added dGeomGetSpace().

02/05/03 russ

	* added dMassSetCylinder().

12/07/02 russ

	* added dAreConnectedExcluding().

11/30/02 russ

	* added the ray geom class.

	* added the dGeomXXXPointDepth() functions.

	* added a collision test infrastructure, and some more tests.

11/24/02 russ

	* added support for multiple box-box contacts.

11/10/02 russ

	* added new collision system. select between the old/new system by
	  setting the ODE_OLD_COLLISION variable in config/user-settings.

10/28/02 russ

	* fixed two problems in the LCP code to improve the reliability of
	  the dContactApprox1 contact mode.

	* added a FAQ question about rolling bodies getting stuck when they
	  hit multiple geoms.

09/08/02 russ

	* added dClosestLineSegmentPoints().
	* implemented dCollideCB().

08/28/02 russ

	* added dJointSetFeedback() and dJointGetFeedback().

08/05/02 russ

	* added dGeomTransformSetInfo() and dGeomTransformGetInfo().

07/13/02 russ


configurator.c
/*************************************************************************
 *                                                                       *
 * Open Dynamics Engine, Copyright (C) 2001,2002 Russell L. Smith.       *
 * All rights reserved.  Email: russ@q12.org   Web: www.q12.org          *
 *                                                                       *
 * This library is free software; you can redistribute it and/or         *
 * modify it under the terms of EITHER:                                  *
 *   (1) The GNU Lesser General Public License as published by the Free  *
 *       Software Foundation; either version 2.1 of the License, or (at  *
 *       your option) any later version. The text of the GNU Lesser      *
 *       General Public License is included with this library in the     *
 *       file LICENSE.TXT.                                               *
 *   (2) The BSD-style license that is included with this library in     *
 *       the file LICENSE-BSD.TXT.                                       *
 *                                                                       *
 * This library is distributed in the hope that it will be useful,       *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the files    *
 * LICENSE.TXT and LICENSE-BSD.TXT for more details.                     *
 *                                                                       *
 *************************************************************************/

/*

this program discovers some system configuration stuff, prior to compiling
ODE. the usage is:

  configurator <config.h-file-to-generate> <compiler-command-line> <THIS_DIR-variable>

this program looks long, but it really has an extremely simple structure and
should be very easy for anyone to modify. it should be very portable as it
is written in straight ANSI C and only uses the following library functions:
  * printf
  * fopen (assumes 0 returned on failure)
  * fclose
  * fprintf
  * system
  * exit
  * unlink
except where stated, we do not assume anything about the return codes from
these functions.

why didn't i just use GNU autoconf? :
  * autoconf needs a bourne shell and a bunch of other tools that windows
    users may not have.
  * i like reinventing the wheel.

*/

/****************************************************************************/
/* include some system header files. <stdio.h> and <stdlib.h> are probably
 * safe on all systems, but <unistd.h> may not be, so we make explicit
 * prototypes for the other functions we need.
 */

#include <stdio.h>
#include <stdlib.h>

#ifndef unlink
#ifdef __cplusplus
extern "C" {
#endif
int unlink (const char *pathname);
#ifdef __cplusplus
}
#endif
#endif

/****************************************************************************/
/* project constants */

/* the config.h header */
char *config_h_part1 =
"/* per-machine configuration. this file is automatically generated. */\n"
"\n"
"#ifndef _ODE_CONFIG_H_\n"
"#define _ODE_CONFIG_H_\n"
"\n"
"/* standard system headers */\n";


char *config_h_part2 =
"\n"
"#ifdef __cplusplus\n"
"extern \"C\" {\n"
"#endif\n"
"\n";

/* the config.h footer */
char *config_h_footer =
"\n"
"#ifdef __cplusplus\n"
"}\n"
"#endif\n"
"#endif\n";

/****************************************************************************/
/* implementations of some string functions. these are prefixed with 'x'
 * to prevent any conflicts with built-in functions.
 */

#define strcpy xstrcpy
void xstrcpy (char *dest, char *src)
{
  while (*src) *dest++ = *src++;
  *dest = 0;
}


#define strcat xstrcat
void xstrcat (char *dest, char *src)
{
  while (*dest) dest++;
  while (*src) *dest++ = *src++;
  *dest = 0;
}

/****************************************************************************/
/* utility functions */

/* print an error message and exit */

void fatal_error (char *message)
{
  printf ("\n*** configurator failed: %s.\n\n"
	  "please fix your configuration and try again.\n"
	  "if you have to fix the configurator program or the makefiles, "
	  "please email\n"
	  "your changes to the ODE mailing list (ode@q12.org).\n\n", message);
  exit (0);
}


/* open a file, generate an error if it can't be done */

FILE * xfopen (char *filename, char *mode)
{
  FILE *f;
  f = fopen (filename,mode);
  if (!f) fatal_error ("can not open a file");
  return f;
}


/* return 1 if the file exists or 0 if not */

int file_exists (char *filename)
{
  FILE *f;
  f = fopen (filename,"rb");
  if (f) fclose (f);
  return (f != 0);
}


/* write a string to a new file */

void write_to_file (char *filename, char *s)
{
  FILE *f = xfopen (filename,"wt");
  fprintf (f,"%s",s);
  fclose (f);
}


/* write a comment to a header file */

void write_header_comment (FILE *file, char *description)
{
  fprintf (file,"/* %s */\n",description);
  printf ("%s ...\n",description);
}


/* delete a file */

void delete_file (char *filename)
{
  unlink (filename);
  if (file_exists (filename)) {
    fatal_error ("the delete_file() function does not work");
  }
}


/* run a compile command */

char *compile_cmd_line = 0;
void compile (char *output, char *input)
{
  char cmd[1000];
  strcpy (cmd,compile_cmd_line);
  strcat (cmd,output);
  strcat (cmd," ");
  strcat (cmd,input);
  printf ("%s\n",cmd);
  if (system (cmd) == -1) {
    printf ("WARNING: may not be able to execute '%s'\n",cmd);
  }
}


/* run a program we've just compiled */

char *run_prefix = "";
void run (char *filename)
{
  char cmd[1000];
  strcpy (cmd,run_prefix);
  strcat (cmd,filename);
  printf ("%s\n",cmd);
  if (system (cmd) == -1) {
    printf ("WARNING: may not be able to execute '%s'\n",cmd);
  }
}

/****************************************************************************/
/* system tests */

void check_if_this_is_a_pentium (FILE *file)
{
  write_header_comment (file,"is this a pentium on a gcc-based platform?");
  write_to_file ("ctest.cpp",
		 "int main() {\n"
		 "  asm (\"mov $0,%%eax\\n cpuid\\n\" : : : \"%eax\");\n"
		 "  return 0;\n"
		 "}\n");
  delete_file ("ctest.exe");
  compile ("ctest.exe","ctest.cpp");
  if (file_exists ("ctest.exe")) {
    fprintf (file,"#define PENTIUM 1\n\n");
  }
  else {
    fprintf (file,"/* #define PENTIUM 1 -- not a pentium */\n\n");
  }

  delete_file ("ctest.cpp");
  delete_file ("ctest.exe");
}


void check_if_this_is_a_64bits (FILE *file)
{
  write_header_comment (file,"is this an x86-64 system on a gcc-based platform?");
  write_to_file ("ctest.cpp",
		 "int main() {\n"
		 "int a = 0; int * pa = &a;\n"
		 "  asm (\"mov %0,%%rax\\n\"\n    \"movl (%%rax),%%eax\"\n"
		 "  : : \"r\"(pa) : \"%rax\");\n"
		 "  return 0;\n"
		 "}\n");
  delete_file ("ctest.exe");
  compile ("ctest.exe","ctest.cpp");
  if (file_exists ("ctest.exe")) {
    fprintf (file,"#define X86_64_SYSTEM 1\n\n");
  }
  else {
    fprintf (file,"/* #define X86_64_SYSTEM 1 -- not an x86-64 system */\n\n");
  }

  delete_file ("ctest.cpp");
  delete_file ("ctest.exe");
}

/****************************************************************************/
/* tests: standard headers */

#define NUM_HEADERS 10
char *header_files[NUM_HEADERS] = {
  "stdio.h", "stdlib.h", "math.h", "string.h",
  "stdarg.h", "malloc.h", "alloca.h",
  "values.h",		// for DBL_MAX and FLT_MAX on some platforms
  "float.h",		// for DBL_MAX and FLT_MAX on some platforms
  "ieeefp.h"		// Solaris needs this apparently
};
int header_used[NUM_HEADERS];


void get_all_standard_headers (FILE *file)
{
  int i;

  for (i=0; i < NUM_HEADERS; i++) {
    FILE *f = xfopen ("ctest.cpp","wt");
    fprintf (f,"#include <%s>\nint main() { return 0; }\n",header_files[i]);
    fclose (f);
    delete_file ("ctest.exe");
    compile ("ctest.exe","ctest.cpp");
    if (file_exists ("ctest.exe")) {
      fprintf (file,"#include <%s>\n",header_files[i]);
      header_used[i] = 1;
    }
    else {
      header_used[i] = 0;
    }
  }

  delete_file ("ctest.cpp");
  delete_file ("ctest.exe");
}

/****************************************************************************/
/* tests: see if alloca() is defined, if not try __builtin_alloca.
 * this assumes that get_all_standard_headers() has been called.
 */

#define NUM_ALLOCA 2
char *alloca_function[NUM_ALLOCA] = {"alloca","__builtin_alloca"};
 
 
void get_alloca_usage (FILE *file)
{
  int i,j;
  for (i=0; i<NUM_ALLOCA; i++) {
    FILE *f = xfopen ("ctest.cpp","wt");
    for (j=0; j < NUM_HEADERS; j++) {
      if (header_used[j]) fprintf (f,"#include <%s>\n",header_files[j]);
    }
    fprintf (f,"int main() { void *foo = %s (10); }\n",alloca_function[i]);
    fclose (f);
    delete_file ("ctest.exe");
    compile ("ctest.exe","ctest.cpp");
    if (file_exists ("ctest.exe")) {
      if (i > 0) {
	write_header_comment (file,"how to use alloca()");
	fprintf (file,"#define alloca %s\n\n",alloca_function[i]);
      }
      break;
    }
  }
  delete_file ("ctest.cpp");
  delete_file ("ctest.exe");
  if (i == NUM_ALLOCA) fatal_error ("i can't find a way to use alloca()");
}

/****************************************************************************/
/* tests: typedefs and constants for ODE */

void get_ODE_integer_typedefs (FILE *file)
{
  write_header_comment (file,"integer types (we assume int >= 32 bits)");
  if (sizeof(char) != 1) fatal_error ("expecting sizeof(char) == 1");
  if (sizeof(int) < 4) fatal_error ("expecting sizeof(int) >= 4");
  fprintf (file,"typedef char int8;\ntypedef unsigned char uint8;\n");

  if (sizeof(short) == 2) {
    fprintf (file,"typedef short int16;\ntypedef unsigned short uint16;\n");
  }
  else {
    fatal_error ("can not find 2 byte integer type");
  }

  if (sizeof(short) == 4) {
    fprintf (file,"typedef short int32;\ntypedef unsigned short uint32;\n");
  }
  else if (sizeof(int) == 4) {
    fprintf (file,"typedef int int32;\ntypedef unsigned int uint32;\n");
  }
  else {
    fatal_error ("can not find 4 byte integer type");
  }

  fprintf (file,"\n"
	   "/* an integer type that we can safely cast a pointer to and\n"
	   " * from without loss of bits.\n"
	   " */\n");
  if (sizeof(short) == sizeof(void*)) {
    fprintf (file,"typedef unsigned short intP;\n");
  }
  else if (sizeof(int) == sizeof(void*)) {
    fprintf (file,"typedef unsigned int intP;\n");
  }
  else if (sizeof(long int) == sizeof(void*)) {
    fprintf (file,"typedef unsigned long int intP;\n");
  }

  fprintf (file,"\n");
}


void get_ODE_float_stuff (FILE *file)
{
  char *suffix;
  int i,j;
  FILE *f;

  #define INFINITIES_TO_TEST 4
#ifdef dSINGLE
  char *decl[INFINITIES_TO_TEST] = {
    "#define dInfinity FLT_MAX",
    "#define dInfinity HUGE_VALF",
    "#define dInfinity 3.402823466e+38F",
    "#define dInfinity 1e20f"};
#else
  char *decl[INFINITIES_TO_TEST] = {
    "#define dInfinity DBL_MAX",
    "#define dInfinity HUGE_VAL",
    "#define dInfinity 1.7976931348623157e+308",
    "#define dInfinity 1e20"};
#endif

  write_header_comment (file,"select the base floating point type");
#ifdef dSINGLE
  fprintf (file,"#define dSINGLE 1\n\n");
  suffix = "f";
#else
  fprintf (file,"#define dDOUBLE 1\n\n");
  suffix = "";
#endif

#ifdef dSINGLE
  fprintf (file,"#define dEpsilon FLT_EPSILON\n\n");
#else
  fprintf (file,"#define dEpsilon DBL_EPSILON\n\n");
#endif

  /* infinity */
  write_header_comment (file,"the floating point infinity");

  /* try the different infinity constants until one works */
  for (i=0; i < INFINITIES_TO_TEST; i++) {
    f = xfopen ("ctest.cpp","wt");
    for (j=0; j < NUM_HEADERS; j++) {
      if (header_used[j]) fprintf (f,"#include <%s>\n",header_files[j]);
    }
    fprintf (f,
	     "%s\n"
	     "int main() {\n"
	     "	if (dInfinity > 1e10%s && -dInfinity < -1e10%s &&\n"
	     "	    -dInfinity < dInfinity) {\n"
	     "	  FILE *f = fopen (\"data\",\"wt\");\n"
	     "	  fprintf (f,\"foo\\n\");\n"
	     "	  fclose (f);\n"
	     "	}\n"
	     "	return 0;\n"
	     "}\n"
	     ,decl[i],suffix,suffix);
    fclose (f);
    delete_file ("data");
    compile ("ctest.exe","ctest.cpp");
    run ("ctest.exe");
    if (file_exists ("data")) {
      fprintf (file,"%s\n\n",decl[i]);
      delete_file ("ctest.cpp");
      delete_file ("ctest.exe");
      delete_file ("data");
      return;
    }
  }

  fatal_error ("can't determine dInfinity constant");
}

/****************************************************************************/
/* configuration of ODE features */

void get_ODE_features (FILE *file)
{
  /* nothing to do here... */
  /* write_header_comment (file,"ODE feature configuration"); */
}

/****************************************************************************
 * figure out what functions are available in the system
 */

#define FUNCTIONS_TO_TEST 10

void get_available_functions (FILE *file)
{
	int i,j;
	char *fnames[FUNCTIONS_TO_TEST*3] = {
		/* statement to compile,	if it works,	if not */
		"sqrtf(1.0f);",			0,		"#define sqrtf sqrt",
		"sinf(1.0f);",			0,		"#define sinf sin",
		"cosf(1.0f);",			0,		"#define cosf cos",
		"fabsf(1.0f);",			0,		"#define fabsf fabs",
		"atan2f(1.0f,1.0f);",		0,		"#define atan2f atan2",
		"fmodf(1.0f,1.0f);",		0,		"#define fmodf fmod",
		"copysignf(1.0f,1.0f);",	0,		"#define copysignf copysign",
		/* to handle cygwin: */
		"copysign(1.0,1.0);",		0,		"#define copysign _copysign",
		"snprintf(\"\",0,\"\");",	0,		"#define snprintf _snprintf",
		"vsnprintf(\"\",0,\"\",0);",	0,		"#define vsnprintf _vsnprintf"
	};

	write_header_comment (file,"available functions");

	for (i=0; i < FUNCTIONS_TO_TEST; i++) {
		FILE *f = xfopen ("ctest.cpp","wt");
		for (j=0; j < NUM_HEADERS; j++) {
			if (header_used[j]) fprintf (f,"#include <%s>\n",header_files[j]);
		}
		fprintf (f,"int main() { %s return 0; }\n",fnames[i*3]);
		fclose (f);
		delete_file ("ctest.exe");
		compile ("ctest.exe","ctest.cpp");
		if (file_exists ("ctest.exe")) {
			if (fnames[i*3+1]) fprintf (file,"%s\n",fnames[i*3+1]);
		}
		else {
			if (fnames[i*3+2]) fprintf (file,"%s\n",fnames[i*3+2]);
		}
		delete_file ("ctest.cpp");
		delete_file ("ctest.exe");
	}
}

/****************************************************************************
 * main
 */

int main (int argc, char **argv)
{
  FILE *file;

  if (argc < 3 || argc > 4)
    fatal_error ("configurator expects 2 or 3 arguments");
  compile_cmd_line = argv[2];
  if (argc >= 4) run_prefix = argv[3];

  printf ("\n*** configurator starting, you may see some harmless error messages ***\n\n");

  /* check some defines we should have been compiled with */
#if !defined(dSINGLE) && !defined(dDOUBLE)
#error you must set the dSINGLE or dDOUBLE macro
  fatal_error ("you must set PRECISION to either SINGLE or DOUBLE");
#endif

  file = xfopen (argv[1],"wt");
  fprintf (file,config_h_part1);
  get_all_standard_headers (file);
  fprintf (file,config_h_part2);
  get_alloca_usage (file);
  check_if_this_is_a_pentium (file);
  check_if_this_is_a_64bits (file);
  get_ODE_integer_typedefs (file);
  get_ODE_float_stuff (file);
  get_ODE_features (file);
  get_available_functions (file);
  fprintf (file,config_h_footer);
  fclose (file);

  printf ("\n*** configurator succeeded ***\n\n");
  return 0;
}
